crates/revm/src/db.rs

```diff
@@ -2,84 +2,91 @@ mod in_memory_db;
 
 #[cfg(feature = "ethersdb")]
 pub mod ethersdb;
-#[cfg(feature = "ethersdb")]
-pub use ethersdb::EthersDB;
 
 #[cfg(all(not(feature = "ethersdb"), feature = "web3db"))]
 compile_error!(
     "`web3db` feature is deprecated, drop-in replacement can be found with feature `ethersdb`"
 );
 
-use crate::AccountInfo;
-use crate::U256;
-use crate::{interpreter::bytecode::Bytecode, Account};
-use crate::{B160, B256};
-use auto_impl::auto_impl;
+#[cfg(feature = "ethersdb")]
+pub use ethersdb::EthersDB;
 use hashbrown::HashMap as Map;
-pub use in_memory_db::{AccountState, BenchmarkDB, CacheDB, DbAccount, EmptyDB, InMemoryDB};
-
-#[auto_impl(& mut, Box)]
-pub trait Database {
-    type Error;
-    /// Get basic account information.
-    fn basic(&mut self, address: B160) -> Result<Option<AccountInfo>, Self::Error>;
-    /// Get account code by its hash
-    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error>;
-    /// Get storage value of address at index.
-    fn storage(&mut self, address: B160, index: U256) -> Result<U256, Self::Error>;
-
-    // History related
-    fn block_hash(&mut self, number: U256) -> Result<B256, Self::Error>;
+use revm_interpreter::{Account, AccountInfo, Bytecode, B160, B256, U256};
+
+use crate::{
+    blockchain::{BlockHash, BlockHashRef},
+    state::{State, StateRef},
+    StateCommit,
+};
+
+pub use self::in_memory_db::{AccountState, BenchmarkDB, CacheDB, DbAccount, EmptyDB, InMemoryDB};
+
+#[impl_tools::autoimpl(for<T: trait> &mut T, Box<T>)]
+pub trait Database: BlockHash + State {
+    type DatabaseError: From<<Self as BlockHash>::Error> + From<<Self as State>::Error>;
```

Comment:
A limitation of this approach it that these constraints will pose problems for `DatabaseComponents` when the two error types are the same